lib/gossip/damper.js

```diff
@@ -0,0 +1,580 @@
+    return findFlappersToDamp(dampScoresByFlapper, rVal, suppressLimit);
+};
+
+Damper.prototype.addDampedMember = function addDampedMember(memberAddr) {
+    this.dampedMembers[memberAddr] = {
+        timestamp: this.Date.now()
+    };
+};
+
+Damper.prototype.addFlapper = function addFlapper(flapper) {
+    if (this.dampedMembers[flapper.address]) {
+        this.logger.debug('ringpop damper already damped member', {
+            local: this.ringpop.whoami()
+        });
+        return false;
+    }
+
+    if (this.flappers[flapper.address]) {
+        this.logger.debug('ringpop damper already added flapper', {
+            local: this.ringpop.whoami()
+        });
+        return false;
+    }
+
+    // TODO Limit number of flappers?
+    // TODO Expire flappers?
+
+    this.flappers[flapper.address] = flapper;
+    this.logger.debug('ringpop damper added flapper', {
+        local: this.ringpop.whoami(),
+        flapper: flapper.address
+    });
+    this.ringpop.stat('increment', 'damper.flapper.added');
+    this.ringpop.stat('gauge', 'damper.flappers',
+        Object.keys(this.flappers).length);
+
+    if (this._getFlapperCount() === 1) {
+        this.logger.debug('ringpop damper starting damp timer; a flapper has been added', {
+            local: this.ringpop.whoami(),
+            flapper: flapper.address
+        });
+        this._startDampTimer();
+    }
+
+    return true;
+};
+
+Damper.prototype.dampMember = function dampMember(addr)  {
+    var member = this.ringpop.membership.findMemberByAddress(addr);
+
+    if (!member) {
+        this.logger.warn('ringpop damper cannot damp member; member does not exist', {
+            local: this.ringpop.whoami(),
+            member: addr
+        });
+        return false;
+    }
+
+    if (!this.hasFlapper(member)) {
+        this.logger.warn('ringpop damper cannot damp member; member is not flappy', {
+            local: this.ringpop.whoami(),
+            member: addr
+        });
+        return false;
+    }
+
+    this.ringpop.membership.makeDamped(member.address);
+
+    // Don't start anymore damping subprotocols for a member
+    // that is already damped.
+    this.removeFlapper(member.address);
+    this.addDampedMember(member.address);
+
+    if (!this.isExpirationTimerEnabled) {
+        this.logger.debug('ringpop damper starting expiration timer; a member has been damped', {
+            local: this.ringpop.whoami(),
+            member: member.address
+        });
+        this._startExpirationTimer();
+    }
+
+    return true;
+};
+
+Damper.prototype.expireDampedMembers = function expireDampedMembers() {
+    var undampedMembers = [];
+    var suppressDuration = this.ringpop.config.get('dampScoringSuppressDuration');
+    var keys = Object.keys(this.dampedMembers);
+    for (var i = 0; i < keys.length; i++) {
+        var key = keys[i];
+        var memberDamping = this.dampedMembers[key];
+        var dampedDuration = this.Date.now() - memberDamping.timestamp;
```

Comment:
this logic could be simplified by doing `var suppressTimestamp = this.Date.now() - suppressDuration;` outside the loop, and then a direct comparison `memberDamping.timestamp >= suppressTimestamp`.